Build a regression model.

In [2]:
# import numpy
import numpy as np
import pandas as pd
import statsmodels.api as sm
import sqlite3

In [5]:
con = sqlite3.connect('../data/ebikes.db') 
query = """
    SELECT *
    FROM ebikes_POI
"""
df = pd.read_sql_query(query, con)
con.close()
df

station_id  station_latitude  station_longitude  \
0     7a19c49f486d7c0c02b3685d7b240448         49.262487        -123.114397   
1     7a19c49f486d7c0c02b3685d7b240448         49.262487        -123.114397   
2     7a19c49f486d7c0c02b3685d7b240448         49.262487        -123.114397   
3     7a19c49f486d7c0c02b3685d7b240448         49.262487        -123.114397   
4     7a19c49f486d7c0c02b3685d7b240448         49.262487        -123.114397   
...                                ...               ...                ...   
2509  e02dd224b0177c65418bcb3f72d2f77d         49.277510        -123.061429   
2510  e02dd224b0177c65418bcb3f72d2f77d         49.277510        -123.061429   
2511  e02dd224b0177c65418bcb3f72d2f77d         49.277510        -123.061429   
2512  e02dd224b0177c65418bcb3f72d2f77d         49.277510        -123.061429   
2513  e02dd224b0177c65418bcb3f72d2f77d         49.277510        -123.061429   

                             POI_name         POI_category  POI_distance  \
0        La Taqueria Pinche Taco Shop      Taco Restaurant           164   
1                         Whole Foods        Grocery Store           182   
2             Hokkaido Ramen Santouka     Ramen Restaurant           186   
3     Cactus Club Cafe Broadway + Ash  American Restaurant           232   
4                     Solly's Bagelry               Bakery           283   
...                               ...                  ...           ...   
2509                  Donald's Market        Grocery Store           501   
2510              Tacofino Commissary   Mexican Restaurant           506   
2511                Platform 7 Coffee                 Café           507   
2512        The Downlow Chicken Shack  Fried Chicken Joint           610   
2513             Kin Kao Thai Kitchen      Thai Restaurant           607   

      free_bikes  
0              3  
1              3  
2              3  
3              3  
4              3  
...          ...  
2509          10  
2510          10  
2511          10  
2512          10  
2513          10  

[2514 rows x 7 columns]

In [6]:
y = df['free_bikes']

In [8]:
x = df['POI_distance']
x = sm.add_constant(x) # adding a constant

In [11]:
lin_reg = sm.OLS(y,x)
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                   0.06063
Date:                Sun, 12 May 2024   Prob (F-statistic):              0.806
Time:                        13:58:01   Log-Likelihood:                -8389.1
No. Observations:                2514   AIC:                         1.678e+04
Df Residuals:                    2512   BIC:                         1.679e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            7.7498      0.288     26.923   

In [18]:
df_group = df.groupby('station_id')[['free_bikes', 'POI_distance']].mean().sort_values(by='POI_distance')
df_group

free_bikes  POI_distance
station_id                                                
ee620d77724c8993b0d366e7ec655b64         2.0          80.7
1d7e73b3321ebb6713774d376247d9f9         4.0         124.0
e75e83499d3fa2e174dc2a8ffd238119        20.0         125.9
f5e8fa1fc4252bda2f93dd27ed2d603f         6.0         132.9
9ee3bc9e836768e3e04e4f3328dae03b         3.0         133.0
...                                      ...           ...
8d2d5bb3b9142f781f2a7f441e6bad7a         5.0         674.9
549f00f6039e8bb702a00dcca08a8ed0         2.0         686.2
899f59a9994302e8e1e4be01117ef707        20.0         696.8
17a3320a52bcc62161908967f8a06613         5.0         726.9
e1ff428dadc7c32141b9d89a7b56728a         6.0         753.3

[252 rows x 2 columns]

In [17]:
y = df_group['free_bikes']
x = df_group['POI_distance']
x = sm.add_constant(x) # adding a constant
lin_reg = sm.OLS(y,x)
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.004
Method:                 Least Squares   F-statistic:                   0.01007
Date:                Sun, 12 May 2024   Prob (F-statistic):              0.920
Time:                        14:04:23   Log-Likelihood:                -841.01
No. Observations:                 252   AIC:                             1686.
Df Residuals:                     250   BIC:                             1693.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            7.6843      1.172      6.558   

Provide model output and an interpretation of the results. 

H0: The points of of interest distance has no significant effect on the number of ebikes available

H1: the points of interest have a significant effect on the number of ebikes available

R squared: is zero so the independent variable (POI distance) doesn't explain the variability in the dependent variable (free ebikes)

P value: pvalue is 0.92, which is much greater that 0.05. A pvalue greater than 0.05 indicates that the obeserved data is consistent with the null hypothesis, which is the POI distance has no significant effect on the number of ebikes available

YELP DATAFRAME
** Note this dataframe has limited data. See yelp_foursquare_EDA for more details

In [19]:
con = sqlite3.connect('../data/yelp.db') 
query = """
    SELECT *
    FROM yelp
"""
yelp = pd.read_sql_query(query, con)
con.close()
yelp

station_id                                  name  \
0   e02dd224b0177c65418bcb3f72d2f77d                        Nonna's Cucina   
1   e02dd224b0177c65418bcb3f72d2f77d                            Lunch Lady   
2   e02dd224b0177c65418bcb3f72d2f77d                         The Red Wagon   
3   e02dd224b0177c65418bcb3f72d2f77d                             Yama Cafe   
4   e02dd224b0177c65418bcb3f72d2f77d                  Bai Bua Thai Cuisine   
5   e02dd224b0177c65418bcb3f72d2f77d                  Royal Palace Seafood   
6   e02dd224b0177c65418bcb3f72d2f77d                         Caffe La Tana   
7   e02dd224b0177c65418bcb3f72d2f77d                                 Dachi   
8   e02dd224b0177c65418bcb3f72d2f77d                   Tacofino Commissary   
9   e02dd224b0177c65418bcb3f72d2f77d                           Sopra Sotto   
10  e02dd224b0177c65418bcb3f72d2f77d                           Mr Red Cafe   
11  e02dd224b0177c65418bcb3f72d2f77d        Jackalope's Neighbourhood Dive   
12  e02dd224b0177c65418bcb3f72d2f77d                   Via Tevere Pizzeria   
13  e02dd224b0177c65418bcb3f72d2f77d                          Aleph Eatery   
14  e02dd224b0177c65418bcb3f72d2f77d                                 Cozen   
15  e02dd224b0177c65418bcb3f72d2f77d                 Enzo’s Rustico Romano   
16  e02dd224b0177c65418bcb3f72d2f77d                    Dundas Eat + Drink   
17  e02dd224b0177c65418bcb3f72d2f77d                       Absinthe Bistro   
18  e02dd224b0177c65418bcb3f72d2f77d                              Pepino's   
19  e02dd224b0177c65418bcb3f72d2f77d                 Downlow Chicken Shack   
20  7a19c49f486d7c0c02b3685d7b240448       Seaport City Seafood Restaurant   
21  7a19c49f486d7c0c02b3685d7b240448                         Marulilu Cafe   
22  7a19c49f486d7c0c02b3685d7b240448                                  Saku   
23  7a19c49f486d7c0c02b3685d7b240448                              Xin Chào   
24  7a19c49f486d7c0c02b3685d7b240448          La Taqueria Pinche Taco Shop   
25  7a19c49f486d7c0c02b3685d7b240448             Pho & Roll by Wooden Fish   
26  7a19c49f486d7c0c02b3685d7b240448                                 Yolks   
27  7a19c49f486d7c0c02b3685d7b240448           ChiuChow Seafood Restaurant   
28  7a19c49f486d7c0c02b3685d7b240448                               A.Bento   
29  7a19c49f486d7c0c02b3685d7b240448                    Osteria Elio Volpe   
30  7a19c49f486d7c0c02b3685d7b240448               Hokkaido Ramen Santouka   
31  7a19c49f486d7c0c02b3685d7b240448  iDen & Quan Ju De Beijing Duck House   
32  7a19c49f486d7c0c02b3685d7b240448                     Mr Japanese Curry   
33  7a19c49f486d7c0c02b3685d7b240448       The Greek By Anatoli - Broadway   
34  7a19c49f486d7c0c02b3685d7b240448                              Alley 16   
35  7a19c49f486d7c0c02b3685d7b240448                                  Nook   
36  7a19c49f486d7c0c02b3685d7b240448                             Uma Sushi   
37  7a19c49f486d7c0c02b3685d7b240448                            Maizal RMF   
38  7a19c49f486d7c0c02b3685d7b240448              Italia Bella Bar & Grill   
39  7a19c49f486d7c0c02b3685d7b240448                    Pho Extreme Xe Lua   

          alias  rating    distance   latitude   longitude  
0       italian     4.7  398.809661  49.280915 -123.060235  
1    vietnamese     4.1  640.141163  49.274970 -123.069150  
2        diners     3.9  413.488794  49.281223 -123.061123  
3         cafes     4.4  458.503706  49.281305 -123.063288  
4          thai     4.1  601.996166  49.281420 -123.055640  
5        dimsum     4.8  440.466744  49.281440 -123.060230  
6         cafes     4.6  680.684474  49.278860 -123.070581  
7     wine_bars     4.0  498.481104  49.281396 -123.058002  
8       mexican     3.9  509.874420  49.281311 -123.057498  
9         pizza     4.0  910.448078  49.271105 -123.069465  
10   vietnamese     3.9  425.174217  49.281317 -123.060882  
11     divebars     4.3  456.399411  49.281285 -123.058654  
12      italian     4.1  502.04524

In [21]:
y = yelp['distance']
x = yelp['rating']
x = sm.add_constant(x) # adding a constant
lin_reg = sm.OLS(y,x)
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:               distance   R-squared:                       0.133
Model:                            OLS   Adj. R-squared:                  0.110
Method:                 Least Squares   F-statistic:                     5.843
Date:                Sun, 12 May 2024   Prob (F-statistic):             0.0205
Time:                        14:24:49   Log-Likelihood:                -276.91
No. Observations:                  40   AIC:                             557.8
Df Residuals:                      38   BIC:                             561.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -52.5362    216.953     -0.242      0.8

H0: There is no significant relationship between POI rating and ebike station distance

H1: There is a significant relationship between POI rating and ebike station distance

R squared: is low at 0.133 so the independent variable (POI rating) doesn't explain the variability in the dependent variable (ebike station distance)

P value: pvalue is 0.021, which is much less than 0.05. A pvalue less than 0.05 indicates we reject the null hypothesis, which is the POI rating has no significant effect on the ebike station distanace

# Stretch

How can you turn the regression model into a classification model?